#### Importing CSV data files as pandas dataframes
For the upcoming exercises, we will use a dataset of a retail store having details about the orders placed, customers, product details, sales, profits etc.

In [1]:
!curl -o global_sales_data/cust_dimen.csv https://raw.githubusercontent.com/sumathi16/Machine-Learning-March/master/data/global_sales_data/cust_dimen.csv

  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed

  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0
  0     0    0     0    0     0      0      0 --:--:--  0:00:01 --:--:--     0
 27  113k   27 31858    0     0  18267      0  0:00:06  0:00:01  0:00:05 18267
100  113k  100  113k    0     0  43727      0  0:00:02  0:00:02 --:--:-- 43727


In [2]:
!curl -o global_sales_data/market_fact.csv https://raw.githubusercontent.com/sumathi16/Machine-Learning-March/master/data/global_sales_data/market_fact.csv

  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed

  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0
  0     0    0     0    0     0      0      0 --:--:--  0:00:01 --:--:--     0
 12  605k   12 76776    0     0  38697      0  0:00:16  0:00:01  0:00:15 38678
 32  605k   32  195k    0     0  67079      0  0:00:09  0:00:02  0:00:07 67057
 55  605k   55  338k    0     0  86911      0  0:00:07  0:00:03  0:00:04 86889
 76  605k   76  465k    0     0  95517      0  0:00:06  0:00:04  0:00:02 95517
 98  605k   98  597k    0     0    99k      0  0:00:06  0:00:06 --:--:--  127k
100  605k  100  605k    0     0    99k      0  0:00:06  0:00:06 --:--:--  129k


In [3]:
!curl -o global_sales_data/orders_dimen.csv https://raw.githubusercontent.com/sumathi16/Machine-Learning-March/master/data/global_sales_data/orders_dimen.csv

  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed

  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0
  0  211k    0  1851    0     0   2134      0  0:01:41 --:--:--  0:01:41  2134
 37  211k   37 81012    0     0  51015      0  0:00:04  0:00:01  0:00:03 51015
 90  211k   90  191k    0     0  75647      0  0:00:02  0:00:02 --:--:-- 75647
100  211k  100  211k    0     0  81512      0  0:00:02  0:00:02 --:--:-- 81512


In [4]:
!curl -o global_sales_data/prod_dimen.csv https://raw.githubusercontent.com/sumathi16/Machine-Learning-March/master/data/global_sales_data/prod_dimen.csv

  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed

  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0
100   797  100   797    0     0    814      0 --:--:-- --:--:-- --:--:--   814
100   797  100   797    0     0    814      0 --:--:-- --:--:-- --:--:--   814


In [5]:
!curl -o global_sales_data/shipping_dimen.csv https://raw.githubusercontent.com/sumathi16/Machine-Learning-March/master/data/global_sales_data/shipping_dimen.csv

  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed

  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0
  8  331k    8 29269    0     0  32960      0  0:00:10 --:--:--  0:00:10 32960
 47  331k   47  159k    0     0  86942      0  0:00:03  0:00:01  0:00:02 86942
100  331k  100  331k    0     0   116k      0  0:00:02  0:00:02 --:--:--  116k


##### reading a CSV file as a dataframe

In [81]:
# read global_sales_data\market_fact.csv into a varaible market_df
import pandas as pd
import numpy as np
market_df = pd.read_csv("global_sales_data\market_fact.csv")
market_df

Ord_id  Prod_id   Ship_id    Cust_id      Sales  Discount  \
0     Ord_5446  Prod_16  SHP_7609  Cust_1818   136.8100      0.01   
1     Ord_5406  Prod_13  SHP_7549  Cust_1818    42.2700      0.01   
2     Ord_5446   Prod_4  SHP_7610  Cust_1818  4701.6900      0.00   
3     Ord_5456   Prod_6  SHP_7625  Cust_1818  2337.8900      0.09   
4     Ord_5485  Prod_17  SHP_7664  Cust_1818  4233.1500      0.08   
...        ...      ...       ...        ...        ...       ...   
8394  Ord_5353   Prod_4  SHP_7479  Cust_1798  2841.4395      0.08   
8395  Ord_5411   Prod_6  SHP_7555  Cust_1798   127.1600      0.10   
8396  Ord_5388   Prod_6  SHP_7524  Cust_1798   243.0500      0.02   
8397  Ord_5348  Prod_15  SHP_7469  Cust_1798  3872.8700      0.03   
8398  Ord_5459   Prod_6  SHP_7628  Cust_1798   603.6900      0.00   

      Order_Quantity   Profit  Shipping_Cost  Product_Base_Margin  
0                 23   -30.51           3.60                 0.56  
1                 13     4.56           0.93                 0.54  
2                 26  1148.90           2.50                 0.59  
3                 43   729.34          14.30                 0.37  
4                 35  1219.87          26.30                 0.38  
...              ...      ...            ...                  ...  
8394              28   374.63           7.69                 0.59  
8395              20   -74.03           6.92                 0.37  
8396              39   -70.85           5.35                 0.40  
8397              23   565.34          30.00                 0.62  
8398              47   131.39           4.86                 0.38  

[8399 rows x 10 columns]

#### Reading and Summarising Dataframes

After you import a dataframe, you'd want to quickly understand its structure, shape, meanings of rows and columns etc. Further, you may want to look at summary statistics - such as mean, percentiles etc.

In [74]:
# Looking at the top and bottom entries of dataframes
market_df.head(3)

Ord_id  Prod_id   Ship_id    Cust_id    Sales  Discount  Order_Quantity  \
0  Ord_5446  Prod_16  SHP_7609  Cust_1818   136.81      0.01              23   
1  Ord_5406  Prod_13  SHP_7549  Cust_1818    42.27      0.01              13   
2  Ord_5446   Prod_4  SHP_7610  Cust_1818  4701.69      0.00              26   

    Profit  Shipping_Cost  Product_Base_Margin  
0   -30.51           3.60                 0.56  
1     4.56           0.93                 0.54  
2  1148.90           2.50                 0.59

In [75]:
market_df.tail(3)

Ord_id  Prod_id   Ship_id    Cust_id    Sales  Discount  \
8396  Ord_5388   Prod_6  SHP_7524  Cust_1798   243.05      0.02   
8397  Ord_5348  Prod_15  SHP_7469  Cust_1798  3872.87      0.03   
8398  Ord_5459   Prod_6  SHP_7628  Cust_1798   603.69      0.00   

      Order_Quantity  Profit  Shipping_Cost  Product_Base_Margin  
8396              39  -70.85           5.35                 0.40  
8397              23  565.34          30.00                 0.62  
8398              47  131.39           4.86                 0.38

In [9]:
market_df.shape

(8399, 10)

Here, each row represents an order placed at a retail store. Notice the index associated with each row - starts at 0 and ends at 8398, implying that there were 8399 orders placed.

In [76]:
# Looking at the datatypes of each column using info
market_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8399 entries, 0 to 8398
Data columns (total 10 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   Ord_id               8399 non-null   object 
 1   Prod_id              8399 non-null   object 
 2   Ship_id              8399 non-null   object 
 3   Cust_id              8399 non-null   object 
 4   Sales                8399 non-null   float64
 5   Discount             8399 non-null   float64
 6   Order_Quantity       8399 non-null   int64  
 7   Profit               8399 non-null   float64
 8   Shipping_Cost        8399 non-null   float64
 9   Product_Base_Margin  8336 non-null   float64
dtypes: float64(5), int64(1), object(4)
memory usage: 656.3+ KB


- Note that each column is basically a pandas Series of length 8399
- The ID columns are 'objects', i.e. they are being read as characters
- The rest are numeric (floats or int)

In [13]:
#Get the summary of all the numeric columns in the dataset
market_df.describe(exclude=object).columns

Index(['Sales', 'Discount', 'Order_Quantity', 'Profit', 'Shipping_Cost',
       'Product_Base_Margin'],
      dtype='object')

In [14]:
# Get the Column names
market_df.columns

Index(['Ord_id', 'Prod_id', 'Ship_id', 'Cust_id', 'Sales', 'Discount',
       'Order_Quantity', 'Profit', 'Shipping_Cost', 'Product_Base_Margin'],
      dtype='object')

In [18]:
# The number of rows and columns
print("rows====",market_df.index)
print("columns====",market_df.columns)

rows==== RangeIndex(start=0, stop=8399, step=1)
columns==== Index(['Ord_id', 'Prod_id', 'Ship_id', 'Cust_id', 'Sales', 'Discount',
       'Order_Quantity', 'Profit', 'Shipping_Cost', 'Product_Base_Margin'],
      dtype='object')


In [22]:
# Extract the values of a dataframe as a numpy array 
market_df.to_numpy

<bound method DataFrame.to_numpy of         Ord_id  Prod_id   Ship_id    Cust_id      Sales  Discount  \
0     Ord_5446  Prod_16  SHP_7609  Cust_1818   136.8100      0.01   
1     Ord_5406  Prod_13  SHP_7549  Cust_1818    42.2700      0.01   
2     Ord_5446   Prod_4  SHP_7610  Cust_1818  4701.6900      0.00   
3     Ord_5456   Prod_6  SHP_7625  Cust_1818  2337.8900      0.09   
4     Ord_5485  Prod_17  SHP_7664  Cust_1818  4233.1500      0.08   
...        ...      ...       ...        ...        ...       ...   
8394  Ord_5353   Prod_4  SHP_7479  Cust_1798  2841.4395      0.08   
8395  Ord_5411   Prod_6  SHP_7555  Cust_1798   127.1600      0.10   
8396  Ord_5388   Prod_6  SHP_7524  Cust_1798   243.0500      0.02   
8397  Ord_5348  Prod_15  SHP_7469  Cust_1798  3872.8700      0.03   
8398  Ord_5459   Prod_6  SHP_7628  Cust_1798   603.6900      0.00   

      Order_Quantity   Profit  Shipping_Cost  Product_Base_Margin  
0                 23   -30.51           3.60                 0.56  

Let's change the index to Ord_id (unique id of each order), so that you can select rows using the order ids directly

In [21]:
# Setting index to Ord_id and store the result in new dataframe market_df_ord
# Don't modify original dataset
market_df_ord = market_df.copy()
market_df_ord = market_df_ord.set_index("Ord_id")
market_df_ord

Prod_id   Ship_id    Cust_id      Sales  Discount  Order_Quantity  \
Ord_id                                                                        
Ord_5446  Prod_16  SHP_7609  Cust_1818   136.8100      0.01              23   
Ord_5406  Prod_13  SHP_7549  Cust_1818    42.2700      0.01              13   
Ord_5446   Prod_4  SHP_7610  Cust_1818  4701.6900      0.00              26   
Ord_5456   Prod_6  SHP_7625  Cust_1818  2337.8900      0.09              43   
Ord_5485  Prod_17  SHP_7664  Cust_1818  4233.1500      0.08              35   
...           ...       ...        ...        ...       ...             ...   
Ord_5353   Prod_4  SHP_7479  Cust_1798  2841.4395      0.08              28   
Ord_5411   Prod_6  SHP_7555  Cust_1798   127.1600      0.10              20   
Ord_5388   Prod_6  SHP_7524  Cust_1798   243.0500      0.02              39   
Ord_5348  Prod_15  SHP_7469  Cust_1798  3872.8700      0.03              23   
Ord_5459   Prod_6  SHP_7628  Cust_1798   603.6900      0.00              47   

           Profit  Shipping_Cost  Product_Base_Margin  
Ord_id                                                 
Ord_5446   -30.51           3.60                 0.56  
Ord_5406     4.56           0.93                 0.54  
Ord_5446  1148.90           2.50                 0.59  
Ord_5456   729.34          14.30                 0.37  
Ord_5485  1219.87          26.30                 0.38  
...           ...            ...                  ...  
Ord_5353   374.63           7.69                 0.59  
Ord_5411   -74.03           6.92                 0.37  
Ord_5388   -70.85           5.35                 0.40  
Ord_5348   565.34          30.00                 0.62  
Ord_5459   131.39           4.86                 0.38  

[8399 rows x 9 columns]

In [25]:
help(market_df_ord.sort_index)

Help on method sort_index in module pandas.core.frame:

sort_index(axis=0, level=None, ascending=True, inplace=False, kind='quicksort', na_position='last', sort_remaining=True, ignore_index:bool=False) method of pandas.core.frame.DataFrame instance
    Sort object by labels (along an axis).
    
    Parameters
    ----------
    axis : {0 or 'index', 1 or 'columns'}, default 0
        The axis along which to sort.  The value 0 identifies the rows,
        and 1 identifies the columns.
    level : int or level name or list of ints or list of level names
        If not None, sort on values in specified index level(s).
    ascending : bool, default True
        Sort ascending vs. descending.
    inplace : bool, default False
        If True, perform operation in-place.
    kind : {'quicksort', 'mergesort', 'heapsort'}, default 'quicksort'
        Choice of sorting algorithm. See also ndarray.np.sort for more
        information.  `mergesort` is the only stable algorithm. For
        DataF

In [26]:
#  Sorting market_df_ord by index
# axis = 0 indicates that you want to sort rows (use axis=1 for columns)
market_df_ord.sort_index(axis=0)

Prod_id   Ship_id   Cust_id     Sales  Discount  Order_Quantity  \
Ord_id                                                                      
Ord_1      Prod_1     SHP_1    Cust_1    261.54      0.04               6   
Ord_10     Prod_3    SHP_13   Cust_10     80.61      0.02              15   
Ord_100    Prod_8   SHP_138   Cust_58    121.12      0.10               3   
Ord_1000   Prod_6  SHP_1384  Cust_373    334.71      0.01              25   
Ord_1001   Prod_5  SHP_1385  Cust_374   1981.26      0.07              49   
...           ...       ...       ...       ...       ...             ...   
Ord_997   Prod_14  SHP_1379  Cust_365  28761.52      0.04               8   
Ord_998    Prod_5  SHP_1382  Cust_372   2149.37      0.03              42   
Ord_998    Prod_8  SHP_1380  Cust_372    750.66      0.00              33   
Ord_998    Prod_8  SHP_1381  Cust_372    254.32      0.01               8   
Ord_999   Prod_15  SHP_1383  Cust_361   5661.08      0.00              33   

           Profit  Shipping_Cost  Product_Base_Margin  
Ord_id                                                 
Ord_1     -213.25          35.00                 0.80  
Ord_10      -4.72           2.99                 0.37  
Ord_100   -118.82           1.99                 0.44  
Ord_1000    31.74           6.47                 0.38  
Ord_1001   100.80           8.66                 0.76  
...           ...            ...                  ...  
Ord_997    285.11          24.49                 0.37  
Ord_998    217.87          19.99                 0.55  
Ord_998    120.05           4.00                 0.60  
Ord_998   -117.39           6.50                 0.79  
Ord_999   1055.47          30.00                 0.62  

[8399 rows x 9 columns]

In [28]:
market_df_ord.sort_index(axis=1)

Cust_id  Discount  Order_Quantity  Prod_id  Product_Base_Margin  \
Ord_id                                                                        
Ord_5446  Cust_1818      0.01              23  Prod_16                 0.56   
Ord_5406  Cust_1818      0.01              13  Prod_13                 0.54   
Ord_5446  Cust_1818      0.00              26   Prod_4                 0.59   
Ord_5456  Cust_1818      0.09              43   Prod_6                 0.37   
Ord_5485  Cust_1818      0.08              35  Prod_17                 0.38   
...             ...       ...             ...      ...                  ...   
Ord_5353  Cust_1798      0.08              28   Prod_4                 0.59   
Ord_5411  Cust_1798      0.10              20   Prod_6                 0.37   
Ord_5388  Cust_1798      0.02              39   Prod_6                 0.40   
Ord_5348  Cust_1798      0.03              23  Prod_15                 0.62   
Ord_5459  Cust_1798      0.00              47   Prod_6                 0.38   

           Profit      Sales   Ship_id  Shipping_Cost  
Ord_id                                                 
Ord_5446   -30.51   136.8100  SHP_7609           3.60  
Ord_5406     4.56    42.2700  SHP_7549           0.93  
Ord_5446  1148.90  4701.6900  SHP_7610           2.50  
Ord_5456   729.34  2337.8900  SHP_7625          14.30  
Ord_5485  1219.87  4233.1500  SHP_7664          26.30  
...           ...        ...       ...            ...  
Ord_5353   374.63  2841.4395  SHP_7479           7.69  
Ord_5411   -74.03   127.1600  SHP_7555           6.92  
Ord_5388   -70.85   243.0500  SHP_7524           5.35  
Ord_5348   565.34  3872.8700  SHP_7469          30.00  
Ord_5459   131.39   603.6900  SHP_7628           4.86  

[8399 rows x 9 columns]

In [29]:
# Sorting market_df_ord by values
# Sorting in increasing order of Sales
market_df_ord.sort_values(by=['Sales'])

Prod_id   Ship_id    Cust_id     Sales  Discount  Order_Quantity  \
Ord_id                                                                       
Ord_704    Prod_7   SHP_964   Cust_242      2.24      0.01               1   
Ord_149    Prod_3  SHP_7028  Cust_1712      3.20      0.09               1   
Ord_4270   Prod_7  SHP_5959  Cust_1450      3.23      0.06               2   
Ord_4755  Prod_13  SHP_6628  Cust_1579      3.41      0.06               1   
Ord_2252   Prod_3  SHP_3064   Cust_881      3.42      0.05               1   
...           ...       ...        ...       ...       ...             ...   
Ord_4614  Prod_14  SHP_6423  Cust_1571  29884.60      0.05              49   
Ord_2373  Prod_14  SHP_3259   Cust_942  33367.85      0.01               9   
Ord_3875  Prod_17  SHP_5370  Cust_1351  41343.21      0.09               8   
Ord_2338  Prod_17  SHP_3207   Cust_932  45923.76      0.07               7   
Ord_3084  Prod_17  SHP_4279  Cust_1151  89061.05      0.00              13   

            Profit  Shipping_Cost  Product_Base_Margin  
Ord_id                                                  
Ord_704      -1.97           0.70                 0.37  
Ord_149      -3.16           1.49                 0.37  
Ord_4270     -2.73           0.70                 0.81  
Ord_4755     -1.78           0.70                 0.56  
Ord_2252     -2.91           1.49                 0.37  
...            ...            ...                  ...  
Ord_4614  12748.86          24.49                 0.44  
Ord_2373   3992.52          24.49                 0.37  
Ord_3875   3852.19          24.49                 0.39  
Ord_2338    102.61          24.49                 0.39  
Ord_3084  27220.69          24.49                 0.39  

[8399 rows x 9 columns]

In [30]:
# Sorting in decreasing order of Shipping_Cost
market_df_ord.sort_values(by='Shipping_Cost',ascending=False)

Prod_id   Ship_id    Cust_id      Sales  Discount  Order_Quantity  \
Ord_id                                                                        
Ord_1751  Prod_15  SHP_2426   Cust_597  14740.510      0.00              46   
Ord_839   Prod_11  SHP_1361   Cust_364  12689.870      0.04              44   
Ord_1741  Prod_11  SHP_2411   Cust_595  15168.820      0.02              26   
Ord_417   Prod_11   SHP_561   Cust_156  20333.816      0.02              45   
Ord_1581  Prod_15  SHP_2184   Cust_519   2573.920      0.07              17   
...           ...       ...        ...        ...       ...             ...   
Ord_4822  Prod_12  SHP_6718  Cust_1654     21.550      0.09               3   
Ord_2375  Prod_12  SHP_3263   Cust_939     90.150      0.07              13   
Ord_3324  Prod_12  SHP_4612  Cust_1178    154.240      0.03              31   
Ord_284   Prod_12   SHP_384   Cust_105     97.130      0.05               1   
Ord_3499  Prod_12  SHP_4846  Cust_1223     86.290      0.02              17   

           Profit  Shipping_Cost  Product_Base_Margin  
Ord_id                                                 
Ord_1751  3407.73         164.73                 0.56  
Ord_839   -169.23         154.12                 0.76  
Ord_1741 -1096.78         147.12                 0.80  
Ord_417  -1430.45         147.12                 0.80  
Ord_1581   117.23         143.71                 0.55  
...           ...            ...                  ...  
Ord_4822    -2.87           0.49                 0.38  
Ord_2375    33.05           0.49                 0.38  
Ord_3324    68.88           0.49                 0.39  
Ord_284    -37.06           0.49                 0.36  
Ord_3499    34.98           0.49                 0.39  

[8399 rows x 9 columns]

In [34]:
# Selecting the rows from indices 2 to 6
market_df_ord.iloc[2:7,:]

Prod_id   Ship_id    Cust_id    Sales  Discount  Order_Quantity  \
Ord_id                                                                      
Ord_5446   Prod_4  SHP_7610  Cust_1818  4701.69      0.00              26   
Ord_5456   Prod_6  SHP_7625  Cust_1818  2337.89      0.09              43   
Ord_5485  Prod_17  SHP_7664  Cust_1818  4233.15      0.08              35   
Ord_5446   Prod_6  SHP_7608  Cust_1818   164.02      0.03              23   
Ord_31    Prod_12    SHP_41    Cust_26    14.76      0.01               5   

           Profit  Shipping_Cost  Product_Base_Margin  
Ord_id                                                 
Ord_5446  1148.90           2.50                 0.59  
Ord_5456   729.34          14.30                 0.37  
Ord_5485  1219.87          26.30                 0.38  
Ord_5446   -47.64           6.15                 0.37  
Ord_31       1.32           0.50                 0.36

In [37]:
# Selecting alternate rows starting from index = 5
market_df_ord.iloc[5::2,:]

Prod_id   Ship_id    Cust_id      Sales  Discount  Order_Quantity  \
Ord_id                                                                        
Ord_5446   Prod_6  SHP_7608  Cust_1818   164.0200      0.03              23   
Ord_4725   Prod_4  SHP_6593  Cust_1641  3410.1575      0.10              48   
Ord_4725   Prod_6  SHP_6593  Cust_1641    57.2200      0.07               8   
Ord_1925   Prod_6  SHP_2637   Cust_708   465.9000      0.05              38   
Ord_2207  Prod_11  SHP_3093   Cust_839  3364.2480      0.10              15   
...           ...       ...        ...        ...       ...             ...   
Ord_4444  Prod_13  SHP_6192  Cust_1519   159.4100      0.00              44   
Ord_5435   Prod_4  SHP_7594  Cust_1798  1991.8985      0.07              20   
Ord_5348   Prod_8  SHP_7470  Cust_1798   356.7200      0.07               9   
Ord_5411   Prod_6  SHP_7555  Cust_1798   127.1600      0.10              20   
Ord_5348  Prod_15  SHP_7469  Cust_1798  3872.8700      0.03              23   

           Profit  Shipping_Cost  Product_Base_Margin  
Ord_id                                                 
Ord_5446   -47.64           6.15                 0.37  
Ord_4725  1137.91           0.99                 0.55  
Ord_4725   -27.72           6.60                 0.37  
Ord_1925    79.34           4.86                 0.38  
Ord_2207  -693.23          61.76                 0.78  
...           ...            ...                  ...  
Ord_4444    34.68           0.98                 0.52  
Ord_5435    88.36           7.69                 0.58  
Ord_5348    12.61           1.99                 0.44  
Ord_5411   -74.03           6.92                 0.37  
Ord_5348   565.34          30.00                 0.62  

[4197 rows x 9 columns]

In [39]:
# Select Sales column as series
market_df_ord.Sales

Ord_id
Ord_5446     136.8100
Ord_5406      42.2700
Ord_5446    4701.6900
Ord_5456    2337.8900
Ord_5485    4233.1500
              ...    
Ord_5353    2841.4395
Ord_5411     127.1600
Ord_5388     243.0500
Ord_5348    3872.8700
Ord_5459     603.6900
Name: Sales, Length: 8399, dtype: float64

In [40]:
# Select Sales column as DataFrame
market_df_ord[["Sales"]]

Sales
Ord_id             
Ord_5446   136.8100
Ord_5406    42.2700
Ord_5446  4701.6900
Ord_5456  2337.8900
Ord_5485  4233.1500
...             ...
Ord_5353  2841.4395
Ord_5411   127.1600
Ord_5388   243.0500
Ord_5348  3872.8700
Ord_5459   603.6900

[8399 rows x 1 columns]

In [42]:
# Select Cust_id, Sales and Profit
market_df_ord[["Cust_id","Sales","Profit"]]

Cust_id      Sales   Profit
Ord_id                                 
Ord_5446  Cust_1818   136.8100   -30.51
Ord_5406  Cust_1818    42.2700     4.56
Ord_5446  Cust_1818  4701.6900  1148.90
Ord_5456  Cust_1818  2337.8900   729.34
Ord_5485  Cust_1818  4233.1500  1219.87
...             ...        ...      ...
Ord_5353  Cust_1798  2841.4395   374.63
Ord_5411  Cust_1798   127.1600   -74.03
Ord_5388  Cust_1798   243.0500   -70.85
Ord_5348  Cust_1798  3872.8700   565.34
Ord_5459  Cust_1798   603.6900   131.39

[8399 rows x 3 columns]

In [262]:
market_df_ord.head(4)

Prod_id   Ship_id    Cust_id    Sales  Discount  Order_Quantity  \
Ord_id                                                                      
Ord_5446  Prod_16  SHP_7609  Cust_1818   136.81      0.01              23   
Ord_5406  Prod_13  SHP_7549  Cust_1818    42.27      0.01              13   
Ord_5446   Prod_4  SHP_7610  Cust_1818  4701.69      0.00              26   
Ord_5456   Prod_6  SHP_7625  Cust_1818  2337.89      0.09              43   

           Profit  Shipping_Cost  Product_Base_Margin  
Ord_id                                                 
Ord_5446   -30.51           3.60                 0.56  
Ord_5406     4.56           0.93                 0.54  
Ord_5446  1148.90           2.50                 0.59  
Ord_5456   729.34          14.30                 0.37

In [263]:
# Selecting a single element
# Select the element in  the third row and fifth column (Sales)
market_df_ord.iloc[3,4]

0.09

In [268]:
# Selecting a single element
# Select row label = 2 and column label = 'Sales
market_df_ord.loc['Ord_5446','Sales']

Ord_id
Ord_5446     136.81
Ord_5446    4701.69
Ord_5446     164.02
Name: Sales, dtype: float64

In [269]:
# Selecting rows using a range of labels from 4 to 8
market_df_ord.iloc[4:9,1:4] #exclude index range

Ship_id    Cust_id      Sales
Ord_id                                  
Ord_5485  SHP_7664  Cust_1818  4233.1500
Ord_5446  SHP_7608  Cust_1818   164.0200
Ord_31      SHP_41    Cust_26    14.7600
Ord_4725  SHP_6593  Cust_1641  3410.1575
Ord_4725  SHP_6593  Cust_1641   162.0000

In [282]:
# selct the row uisng loc and iloc and observe  difference between iloc and loc
market_df_ord.loc[["Ord_5446"],:] #include index range

Prod_id   Ship_id    Cust_id    Sales  Discount  Order_Quantity  \
Ord_id                                                                      
Ord_5446  Prod_16  SHP_7609  Cust_1818   136.81      0.01              23   
Ord_5446   Prod_4  SHP_7610  Cust_1818  4701.69      0.00              26   
Ord_5446   Prod_6  SHP_7608  Cust_1818   164.02      0.03              23   

           Profit  Shipping_Cost  Product_Base_Margin  
Ord_id                                                 
Ord_5446   -30.51           3.60                 0.56  
Ord_5446  1148.90           2.50                 0.59  
Ord_5446   -47.64           6.15                 0.37

In [287]:
# Select Ord_id = Ord_5406 and 'Sales', 'Profit', 'Cust_id' columns
df1 = market_df_ord[["Sales","Profit","Cust_id"]]
#df1.loc[df1['Ord_id'] == 'Ord_5406']

In [291]:
df1.loc['Ord_5406']

Sales          42.27
Profit          4.56
Cust_id    Cust_1818
Name: Ord_5406, dtype: object

In [292]:
# Select all rows where Sales > 3000
market_df_ord[market_df_ord["Sales"]>3000]

Prod_id   Ship_id    Cust_id       Sales  Discount  Order_Quantity  \
Ord_id                                                                         
Ord_5446   Prod_4  SHP_7610  Cust_1818   4701.6900      0.00              26   
Ord_5485  Prod_17  SHP_7664  Cust_1818   4233.1500      0.08              35   
Ord_4725   Prod_4  SHP_6593  Cust_1641   3410.1575      0.10              48   
Ord_4743   Prod_2  SHP_6615  Cust_1641   4072.0100      0.01              43   
Ord_2207  Prod_11  SHP_3093   Cust_839   3364.2480      0.10              15   
...           ...       ...        ...         ...       ...             ...   
Ord_3593   Prod_3  SHP_4974  Cust_1274  12073.0600      0.03              39   
Ord_3593  Prod_15  SHP_4975  Cust_1274   6685.0500      0.09              25   
Ord_2624   Prod_4  SHP_3591  Cust_1006   4924.1350      0.07              28   
Ord_2722   Prod_1  SHP_3731  Cust_1006   3508.3300      0.04              21   
Ord_5348  Prod_15  SHP_7469  Cust_1798   3872.8700      0.03              23   

           Profit  Shipping_Cost  Product_Base_Margin  
Ord_id                                                 
Ord_5446  1148.90           2.50                 0.59  
Ord_5485  1219.87          26.30                 0.38  
Ord_4725  1137.91           0.99                 0.55  
Ord_4743  1675.98           0.99                 0.56  
Ord_2207  -693.23          61.76                 0.78  
...           ...            ...                  ...  
Ord_3593  5081.87          19.99                 0.38  
Ord_3593  1653.60          24.49                  NaN  
Ord_2624  1049.54           8.99                 0.58  
Ord_2722  -546.98          35.00                 0.85  
Ord_5348   565.34          30.00                 0.62  

[1359 rows x 9 columns]

In [293]:
# First, we get a boolean array where True corresponds to rows having Sales > 3000
market_df_ord["Sales"]>3000

Ord_id
Ord_5446    False
Ord_5406    False
Ord_5446     True
Ord_5456    False
Ord_5485     True
            ...  
Ord_5353    False
Ord_5411    False
Ord_5388    False
Ord_5348     True
Ord_5459    False
Name: Sales, Length: 8399, dtype: bool

In [294]:
# Then, we pass this boolean array inside df.loc
market_df_ord.loc[(market_df_ord["Sales"]>3000)]

Prod_id   Ship_id    Cust_id       Sales  Discount  Order_Quantity  \
Ord_id                                                                         
Ord_5446   Prod_4  SHP_7610  Cust_1818   4701.6900      0.00              26   
Ord_5485  Prod_17  SHP_7664  Cust_1818   4233.1500      0.08              35   
Ord_4725   Prod_4  SHP_6593  Cust_1641   3410.1575      0.10              48   
Ord_4743   Prod_2  SHP_6615  Cust_1641   4072.0100      0.01              43   
Ord_2207  Prod_11  SHP_3093   Cust_839   3364.2480      0.10              15   
...           ...       ...        ...         ...       ...             ...   
Ord_3593   Prod_3  SHP_4974  Cust_1274  12073.0600      0.03              39   
Ord_3593  Prod_15  SHP_4975  Cust_1274   6685.0500      0.09              25   
Ord_2624   Prod_4  SHP_3591  Cust_1006   4924.1350      0.07              28   
Ord_2722   Prod_1  SHP_3731  Cust_1006   3508.3300      0.04              21   
Ord_5348  Prod_15  SHP_7469  Cust_1798   3872.8700      0.03              23   

           Profit  Shipping_Cost  Product_Base_Margin  
Ord_id                                                 
Ord_5446  1148.90           2.50                 0.59  
Ord_5485  1219.87          26.30                 0.38  
Ord_4725  1137.91           0.99                 0.55  
Ord_4743  1675.98           0.99                 0.56  
Ord_2207  -693.23          61.76                 0.78  
...           ...            ...                  ...  
Ord_3593  5081.87          19.99                 0.38  
Ord_3593  1653.60          24.49                  NaN  
Ord_2624  1049.54           8.99                 0.58  
Ord_2722  -546.98          35.00                 0.85  
Ord_5348   565.34          30.00                 0.62  

[1359 rows x 9 columns]

In [295]:
#Select  all orders having 2000 < Sales < 3000 and Profit > 100
market_df_ord.loc[(market_df_ord["Sales"]>2000)&(market_df_ord["Sales"]<3000)&(market_df_ord["Profit"]>100)]

Prod_id   Ship_id    Cust_id      Sales  Discount  Order_Quantity  \
Ord_id                                                                        
Ord_5456   Prod_6  SHP_7625  Cust_1818  2337.8900      0.09              43   
Ord_5205   Prod_4  SHP_7274  Cust_1749  2546.5235      0.09              26   
Ord_139   Prod_17   SHP_186    Cust_45  2671.2100      0.06              14   
Ord_239    Prod_4   SHP_332    Cust_45  2157.3085      0.00              38   
Ord_1673  Prod_17  SHP_2314   Cust_498  2027.5500      0.04              14   
...           ...       ...        ...        ...       ...             ...   
Ord_2107   Prod_2  SHP_2882   Cust_785  2409.9600      0.07              32   
Ord_3570   Prod_4  SHP_4942  Cust_1266  2094.9780      0.06              44   
Ord_3592   Prod_4  SHP_4973  Cust_1266  2614.3705      0.07              25   
Ord_2696   Prod_4  SHP_3691  Cust_1006  2836.0505      0.01              25   
Ord_5353   Prod_4  SHP_7479  Cust_1798  2841.4395      0.08              28   

          Profit  Shipping_Cost  Product_Base_Margin  
Ord_id                                                
Ord_5456  729.34          14.30                 0.37  
Ord_5205  210.00           7.69                 0.59  
Ord_139   636.18          15.59                 0.36  
Ord_239   519.25           5.31                 0.57  
Ord_1673  537.40          13.99                 0.37  
...          ...            ...                  ...  
Ord_2107  575.10           4.50                 0.59  
Ord_3570  697.29           1.25                 0.55  
Ord_3592  384.01           7.69                 0.58  
Ord_2696  561.13           8.99                 0.59  
Ord_5353  374.63           7.69                 0.59  

[328 rows x 9 columns]

In [300]:
#Select all orders having 2000 < Sales  OR Profit > 100
market_df_ord.loc[(market_df_ord["Sales"]<2000)|(market_df_ord["Profit"]>100)]

Prod_id   Ship_id    Cust_id      Sales  Discount  Order_Quantity  \
Ord_id                                                                        
Ord_5446  Prod_16  SHP_7609  Cust_1818   136.8100      0.01              23   
Ord_5406  Prod_13  SHP_7549  Cust_1818    42.2700      0.01              13   
Ord_5446   Prod_4  SHP_7610  Cust_1818  4701.6900      0.00              26   
Ord_5456   Prod_6  SHP_7625  Cust_1818  2337.8900      0.09              43   
Ord_5485  Prod_17  SHP_7664  Cust_1818  4233.1500      0.08              35   
...           ...       ...        ...        ...       ...             ...   
Ord_5353   Prod_4  SHP_7479  Cust_1798  2841.4395      0.08              28   
Ord_5411   Prod_6  SHP_7555  Cust_1798   127.1600      0.10              20   
Ord_5388   Prod_6  SHP_7524  Cust_1798   243.0500      0.02              39   
Ord_5348  Prod_15  SHP_7469  Cust_1798  3872.8700      0.03              23   
Ord_5459   Prod_6  SHP_7628  Cust_1798   603.6900      0.00              47   

           Profit  Shipping_Cost  Product_Base_Margin  
Ord_id                                                 
Ord_5446   -30.51           3.60                 0.56  
Ord_5406     4.56           0.93                 0.54  
Ord_5446  1148.90           2.50                 0.59  
Ord_5456   729.34          14.30                 0.37  
Ord_5485  1219.87          26.30                 0.38  
...           ...            ...                  ...  
Ord_5353   374.63           7.69                 0.59  
Ord_5411   -74.03           6.92                 0.37  
Ord_5388   -70.85           5.35                 0.40  
Ord_5348   565.34          30.00                 0.62  
Ord_5459   131.39           4.86                 0.38  

[7822 rows x 9 columns]

In [296]:
# Select all orders having 2000 < Sales < 3000 and Profit > 100
# Also, this time, you only need the Cust_id, Sales and Profit columns
df2 = market_df_ord.loc[(market_df_ord["Sales"]>2000)&(market_df_ord["Sales"]<3000)&(market_df_ord["Profit"]>100)]
df2[["Cust_id","Sales","Profit"]]

Cust_id      Sales  Profit
Ord_id                                
Ord_5456  Cust_1818  2337.8900  729.34
Ord_5205  Cust_1749  2546.5235  210.00
Ord_139     Cust_45  2671.2100  636.18
Ord_239     Cust_45  2157.3085  519.25
Ord_1673   Cust_498  2027.5500  537.40
...             ...        ...     ...
Ord_2107   Cust_785  2409.9600  575.10
Ord_3570  Cust_1266  2094.9780  697.29
Ord_3592  Cust_1266  2614.3705  384.01
Ord_2696  Cust_1006  2836.0505  561.13
Ord_5353  Cust_1798  2841.4395  374.63

[328 rows x 3 columns]

In [301]:
market_df_ord.columns

Index(['Prod_id', 'Ship_id', 'Cust_id', 'Sales', 'Discount', 'Order_Quantity',
       'Profit', 'Shipping_Cost', 'Product_Base_Margin'],
      dtype='object')

In [303]:
# Get the orders where sales is equal to 4233.15
market_df_ord.loc[(market_df_ord["Sales"]==4233.15)]

Prod_id   Ship_id    Cust_id    Sales  Discount  Order_Quantity  \
Ord_id                                                                      
Ord_5485  Prod_17  SHP_7664  Cust_1818  4233.15      0.08              35   

           Profit  Shipping_Cost  Product_Base_Margin  
Ord_id                                                 
Ord_5485  1219.87           26.3                 0.38

In [304]:
market_df_ord.loc[(market_df_ord["Cust_id"]=='Cust_1798')]

Prod_id   Ship_id    Cust_id      Sales  Discount  Order_Quantity  \
Ord_id                                                                        
Ord_5435  Prod_16  SHP_7594  Cust_1798   316.9900      0.04              47   
Ord_5435   Prod_4  SHP_7594  Cust_1798  1991.8985      0.07              20   
Ord_5384   Prod_9  SHP_7519  Cust_1798   181.5000      0.08              43   
Ord_5348   Prod_8  SHP_7470  Cust_1798   356.7200      0.07               9   
Ord_5353   Prod_4  SHP_7479  Cust_1798  2841.4395      0.08              28   
Ord_5411   Prod_6  SHP_7555  Cust_1798   127.1600      0.10              20   
Ord_5388   Prod_6  SHP_7524  Cust_1798   243.0500      0.02              39   
Ord_5348  Prod_15  SHP_7469  Cust_1798  3872.8700      0.03              23   
Ord_5459   Prod_6  SHP_7628  Cust_1798   603.6900      0.00              47   

          Profit  Shipping_Cost  Product_Base_Margin  
Ord_id                                                
Ord_5435 -276.54           8.37                 0.58  
Ord_5435   88.36           7.69                 0.58  
Ord_5384   -6.24           2.50                 0.37  
Ord_5348   12.61           1.99                 0.44  
Ord_5353  374.63           7.69                 0.59  
Ord_5411  -74.03           6.92                 0.37  
Ord_5388  -70.85           5.35                 0.40  
Ord_5348  565.34          30.00                 0.62  
Ord_5459  131.39           4.86                 0.38

In [305]:
# For instance, say a colleague gives you a list of customer_ids from a certain region

customers_in_hyd = ['Cust_1798', 'Cust_1519', 'Cust_637', 'Cust_851']

# get all the orders from these customers
market_df_ord[market_df_ord.Cust_id.isin(customers_in_hyd)]

Prod_id   Ship_id    Cust_id      Sales  Discount  Order_Quantity  \
Ord_id                                                                        
Ord_1833   Prod_3  SHP_2527   Cust_637   611.1600      0.04              46   
Ord_2324   Prod_7  SHP_3189   Cust_851   121.8700      0.07              39   
Ord_2220   Prod_3  SHP_3019   Cust_851    41.0600      0.04               4   
Ord_4424   Prod_1  SHP_6165  Cust_1519   994.0400      0.03              10   
Ord_4444  Prod_13  SHP_6192  Cust_1519   159.4100      0.00              44   
Ord_5435  Prod_16  SHP_7594  Cust_1798   316.9900      0.04              47   
Ord_5435   Prod_4  SHP_7594  Cust_1798  1991.8985      0.07              20   
Ord_5384   Prod_9  SHP_7519  Cust_1798   181.5000      0.08              43   
Ord_5348   Prod_8  SHP_7470  Cust_1798   356.7200      0.07               9   
Ord_5353   Prod_4  SHP_7479  Cust_1798  2841.4395      0.08              28   
Ord_5411   Prod_6  SHP_7555  Cust_1798   127.1600      0.10              20   
Ord_5388   Prod_6  SHP_7524  Cust_1798   243.0500      0.02              39   
Ord_5348  Prod_15  SHP_7469  Cust_1798  3872.8700      0.03              23   
Ord_5459   Prod_6  SHP_7628  Cust_1798   603.6900      0.00              47   

          Profit  Shipping_Cost  Product_Base_Margin  
Ord_id                                                
Ord_1833  100.22           4.98                 0.40  
Ord_2324   11.32           1.35                 0.40  
Ord_2220  -16.39           6.28                 0.35  
Ord_4424 -335.06          35.00                  NaN  
Ord_4444   34.68           0.98                 0.52  
Ord_5435 -276.54           8.37                 0.58  
Ord_5435   88.36           7.69                 0.58  
Ord_5384   -6.24           2.50                 0.37  
Ord_5348   12.61           1.99                 0.44  
Ord_5353  374.63           7.69                 0.59  
Ord_5411  -74.03           6.92                 0.37  
Ord_5388  -70.85           5.35                 0.40  
Ord_5348  565.34          30.00                 0.62  
Ord_5459  131.39           4.86                 0.38

In [137]:
# Read the data from "global_sales_data/cust_dimen.csv" into customer_df
customer_df = pd.read_csv("global_sales_data/cust_dimen.csv")
customer_df

Customer_Name Province   Region Customer_Segment    Cust_id
0     MUHAMMED MACINTYRE  NUNAVUT  NUNAVUT   SMALL BUSINESS     Cust_1
1           BARRY FRENCH  NUNAVUT  NUNAVUT         CONSUMER     Cust_2
2          CLAY ROZENDAL  NUNAVUT  NUNAVUT        CORPORATE     Cust_3
3         CARLOS SOLTERO  NUNAVUT  NUNAVUT         CONSUMER     Cust_4
4           CARL JACKSON  NUNAVUT  NUNAVUT        CORPORATE     Cust_5
...                  ...      ...      ...              ...        ...
1827      NICOLE BRENNAN  ALBERTA     WEST         CONSUMER  Cust_1828
1828       JASON FORTUNE  ALBERTA     WEST        CORPORATE  Cust_1829
1829        HARRY GREENE  ALBERTA     WEST        CORPORATE  Cust_1830
1830     GRANT DONATELLI  ALBERTA     WEST         CONSUMER  Cust_1831
1831          MICK BROWN  ALBERTA     WEST         CONSUMER  Cust_1832

[1832 rows x 5 columns]

In [142]:
# Customer dimension table: Each row contains metadata about customers
# Display the first five row
print(customer_df.shape)
print(customer_df.head(5))

(1832, 5)
        Customer_Name Province   Region Customer_Segment Cust_id
0  MUHAMMED MACINTYRE  NUNAVUT  NUNAVUT   SMALL BUSINESS  Cust_1
1        BARRY FRENCH  NUNAVUT  NUNAVUT         CONSUMER  Cust_2
2       CLAY ROZENDAL  NUNAVUT  NUNAVUT        CORPORATE  Cust_3
3      CARLOS SOLTERO  NUNAVUT  NUNAVUT         CONSUMER  Cust_4
4        CARL JACKSON  NUNAVUT  NUNAVUT        CORPORATE  Cust_5


In [138]:
# Read the data from "global_sales_data/orders_dimen.csv.csv" into orders_df
orders_df = pd.read_csv("global_sales_data/orders_dimen.csv")
orders_df

Order_ID  Order_Date Order_Priority    Ord_id
0            3  13-10-2010            LOW     Ord_1
1          293  01-10-2012           HIGH     Ord_2
2          483  10-07-2011           HIGH     Ord_3
3          515  28-08-2010  NOT SPECIFIED     Ord_4
4          613  17-06-2011           HIGH     Ord_5
...        ...         ...            ...       ...
5501     47360  08-10-2010  NOT SPECIFIED  Ord_5502
5502     52706  09-07-2012            LOW  Ord_5503
5503     54279  30-07-2011           HIGH  Ord_5504
5504     55558  09-08-2010         MEDIUM  Ord_5505
5505     56581  08-02-2009           HIGH  Ord_5506

[5506 rows x 4 columns]

In [146]:
# Orders dimension table
# Display the first five row
print(orders_df.head())
print(orders_df.shape)

   Order_ID  Order_Date Order_Priority Ord_id
0         3  13-10-2010            LOW  Ord_1
1       293  01-10-2012           HIGH  Ord_2
2       483  10-07-2011           HIGH  Ord_3
3       515  28-08-2010  NOT SPECIFIED  Ord_4
4       613  17-06-2011           HIGH  Ord_5
(5506, 4)


In [139]:
# Read the data from "global_sales_data/prod_dimen.csv" into product_df
product_df = pd.read_csv("global_sales_data/prod_dimen.csv")
product_df

Product_Category            Product_Sub_Category  Prod_id
0   OFFICE SUPPLIES          STORAGE & ORGANIZATION   Prod_1
1   OFFICE SUPPLIES                      APPLIANCES   Prod_2
2   OFFICE SUPPLIES  BINDERS AND BINDER ACCESSORIES   Prod_3
3        TECHNOLOGY    TELEPHONES AND COMMUNICATION   Prod_4
4         FURNITURE              OFFICE FURNISHINGS   Prod_5
5   OFFICE SUPPLIES                           PAPER   Prod_6
6   OFFICE SUPPLIES                    RUBBER BANDS   Prod_7
7        TECHNOLOGY            COMPUTER PERIPHERALS   Prod_8
8   OFFICE SUPPLIES                       ENVELOPES   Prod_9
9         FURNITURE                       BOOKCASES  Prod_10
10        FURNITURE                          TABLES  Prod_11
11  OFFICE SUPPLIES                          LABELS  Prod_12
12  OFFICE SUPPLIES             PENS & ART SUPPLIES  Prod_13
13       TECHNOLOGY                 COPIERS AND FAX  Prod_14
14        FURNITURE              CHAIRS & CHAIRMATS  Prod_15
15  OFFICE SUPPLIES   SCISSORS, RULERS AND TRIMMERS  Prod_16
16       TECHNOLOGY                 OFFICE MACHINES  Prod_17

In [144]:
# Product dimension table
# Display the first five row
print(product_df.ndim)
print(product_df.shape)
print(product_df.head())

2
(17, 3)
  Product_Category            Product_Sub_Category Prod_id
0  OFFICE SUPPLIES          STORAGE & ORGANIZATION  Prod_1
1  OFFICE SUPPLIES                      APPLIANCES  Prod_2
2  OFFICE SUPPLIES  BINDERS AND BINDER ACCESSORIES  Prod_3
3       TECHNOLOGY    TELEPHONES AND COMMUNICATION  Prod_4
4        FURNITURE              OFFICE FURNISHINGS  Prod_5


In [141]:
# Read the data from "global_sales_data/shipping_dimen.csv" into shipping_df
shipping_df = pd.read_csv("global_sales_data/shipping_dimen.csv")
shipping_df

Order_ID       Ship_Mode   Ship_Date   Ship_id
0            3     REGULAR AIR  20-10-2010     SHP_1
1          293  DELIVERY TRUCK  02-10-2012     SHP_2
2          293     REGULAR AIR  03-10-2012     SHP_3
3          483     REGULAR AIR  12-07-2011     SHP_4
4          515     REGULAR AIR  30-08-2010     SHP_5
...        ...             ...         ...       ...
7696     55558  DELIVERY TRUCK  09-08-2010  SHP_7697
7697     55558     REGULAR AIR  11-08-2010  SHP_7698
7698     56550     EXPRESS AIR  10-04-2011  SHP_7699
7699     56550     REGULAR AIR  09-04-2011  SHP_7700
7700     56581     EXPRESS AIR  11-02-2009  SHP_7701

[7701 rows x 4 columns]

In [147]:
# Shipping metadata
# Display the first five row
print(shipping_df.shape)
print(shipping_df.head())

(7701, 4)
   Order_ID       Ship_Mode   Ship_Date Ship_id
0         3     REGULAR AIR  20-10-2010   SHP_1
1       293  DELIVERY TRUCK  02-10-2012   SHP_2
2       293     REGULAR AIR  03-10-2012   SHP_3
3       483     REGULAR AIR  12-07-2011   SHP_4
4       515     REGULAR AIR  30-08-2010   SHP_5


In [149]:
help(pd.merge)

Help on function merge in module pandas.core.reshape.merge:

merge(left, right, how:str='inner', on=None, left_on=None, right_on=None, left_index:bool=False, right_index:bool=False, sort:bool=False, suffixes=('_x', '_y'), copy:bool=True, indicator:bool=False, validate=None) -> 'DataFrame'
    Merge DataFrame or named Series objects with a database-style join.
    
    The join is done on columns or indexes. If joining columns on
    columns, the DataFrame indexes *will be ignored*. Otherwise if joining indexes
    on indexes or indexes on a column or columns, the index will be passed on.
    
    Parameters
    ----------
    left : DataFrame
    right : DataFrame or named Series
        Object to merge with.
    how : {'left', 'right', 'outer', 'inner'}, default 'inner'
        Type of merge to be performed.
    
        * left: use only keys from left frame, similar to a SQL left outer join;
          preserve key order.
        * right: use only keys from right frame, similar to a S

In [150]:
market_df.shape

(8399, 10)

In [156]:
market_df.columns

Index(['Ord_id', 'Prod_id', 'Ship_id', 'Cust_id', 'Sales', 'Discount',
       'Order_Quantity', 'Profit', 'Shipping_Cost', 'Product_Base_Margin'],
      dtype='object')

In [151]:
customer_df.shape

(1832, 5)

In [157]:
customer_df.columns

Index(['Customer_Name', 'Province', 'Region', 'Customer_Segment', 'Cust_id'], dtype='object')

In [315]:
# Merging the dataframes
# Note that Cust_id is the common column/key, in market_df and customer_df
# Make sure the  customer ids present in both dfs are included in the result df1
df1 = pd.merge(market_df,customer_df)
df1.shape

(8399, 14)

In [169]:
df1.head()

Customer_Name Province   Region Customer_Segment Cust_id  Ord_id  \
0  MUHAMMED MACINTYRE  NUNAVUT  NUNAVUT   SMALL BUSINESS  Cust_1   Ord_1   
1        BARRY FRENCH  NUNAVUT  NUNAVUT         CONSUMER  Cust_2   Ord_2   
2        BARRY FRENCH  NUNAVUT  NUNAVUT         CONSUMER  Cust_2   Ord_2   
3       CLAY ROZENDAL  NUNAVUT  NUNAVUT        CORPORATE  Cust_3   Ord_3   
4       CLAY ROZENDAL  NUNAVUT  NUNAVUT        CORPORATE  Cust_3  Ord_35   

  Prod_id Ship_id       Sales  Discount  Order_Quantity   Profit  \
0  Prod_1   SHP_1    261.5400      0.04               6  -213.25   
1  Prod_3   SHP_3    244.5700      0.01              27    46.71   
2  Prod_2   SHP_2  10123.0200      0.07              49   457.81   
3  Prod_4   SHP_4   4965.7595      0.08              30  1198.97   
4  Prod_8  SHP_47   1285.3700      0.10              42   514.07   

   Shipping_Cost  Product_Base_Margin  
0          35.00                 0.80  
1           2.99                 0.39  
2          68.02                 0.58  
3           3.99                 0.58  
4           1.99                 0.42

In [171]:
# Now, you can subset the orders made by customers from 'Corporate' segment in df1
customer_segment = df1.loc[(df1["Customer_Segment"]=='CORPORATE')]
customer_segment

Customer_Name Province   Region Customer_Segment    Cust_id    Ord_id  \
3      CLAY ROZENDAL  NUNAVUT  NUNAVUT        CORPORATE     Cust_3     Ord_3   
4      CLAY ROZENDAL  NUNAVUT  NUNAVUT        CORPORATE     Cust_3    Ord_35   
10      CARL JACKSON  NUNAVUT  NUNAVUT        CORPORATE     Cust_5     Ord_5   
11      CARL JACKSON  NUNAVUT  NUNAVUT        CORPORATE     Cust_5     Ord_5   
12    MONICA FEDERLE  NUNAVUT  NUNAVUT        CORPORATE     Cust_6     Ord_6   
...              ...      ...      ...              ...        ...       ...   
8389   JASON FORTUNE  ALBERTA     WEST        CORPORATE  Cust_1829  Ord_3024   
8390   JASON FORTUNE  ALBERTA     WEST        CORPORATE  Cust_1829  Ord_5499   
8391    HARRY GREENE  ALBERTA     WEST        CORPORATE  Cust_1830  Ord_5500   
8392    HARRY GREENE  ALBERTA     WEST        CORPORATE  Cust_1830  Ord_5504   
8393    HARRY GREENE  ALBERTA     WEST        CORPORATE  Cust_1830  Ord_5503   

      Prod_id   Ship_id       Sales  Discount  Order_Quantity   Profit  \
3      Prod_4     SHP_4   4965.7595      0.08              30  1198.97   
4      Prod_8    SHP_47   1285.3700      0.10              42   514.07   
10     Prod_1     SHP_7    905.0800      0.09              22   127.70   
11     Prod_3     SHP_6     93.5400      0.03              12   -54.04   
12     Prod_1     SHP_8   2781.8200      0.07              21  -695.26   
...       ...       ...         ...       ...             ...      ...   
8389   Prod_4  SHP_7691   5347.9875      0.01              31  1332.44   
8390   Prod_4  SHP_7688   5126.4180      0.03              47  1557.64   
8391   Prod_3  SHP_7689     82.1200      0.07              23  -121.59   
8392  Prod_17  SHP_7696  10071.0900      0.10              41  1977.69   
8393   Prod_3  SHP_7695   1041.6600      0.02              34   480.53   

      Shipping_Cost  Product_Base_Margin  
3              3.99                 0.58  
4              1.99                 0.42  
10             6.22                  NaN  
11             7.72                 0.38  
12            35.00                  NaN  
...             ...                  ...  
8389           4.20                 0.59  
8390           2.50                 0.60  
8391           6.83                 0.38  
8392          17.86                 0.58  
8393           1.49                 0.38  

[3076 rows x 14 columns]

In [173]:
#  merge the df1 and product_df  store the result in df2
df2 = pd.merge(df1,product_df)
df2

Customer_Name Province   Region Customer_Segment    Cust_id  \
0     MUHAMMED MACINTYRE  NUNAVUT  NUNAVUT   SMALL BUSINESS     Cust_1   
1         CARLOS SOLTERO  NUNAVUT  NUNAVUT         CONSUMER     Cust_4   
2           CARL JACKSON  NUNAVUT  NUNAVUT        CORPORATE     Cust_5   
3         MONICA FEDERLE  NUNAVUT  NUNAVUT        CORPORATE     Cust_6   
4          NICOLE HANSEN  NUNAVUT  NUNAVUT   SMALL BUSINESS    Cust_24   
...                  ...      ...      ...              ...        ...   
8394            SHUI TOM  ALBERTA     WEST      HOME OFFICE  Cust_1816   
8395        FRANK HAWLEY  ALBERTA     WEST      HOME OFFICE  Cust_1817   
8396       AARON BERGMAN  ALBERTA     WEST        CORPORATE  Cust_1818   
8397        ADRIAN SHAMI  ALBERTA     WEST         CONSUMER  Cust_1820   
8398        HARRY GREENE  ALBERTA     WEST        CORPORATE  Cust_1830   

        Ord_id  Prod_id   Ship_id     Sales  Discount  Order_Quantity  \
0        Ord_1   Prod_1     SHP_1    261.54      0.04               6   
1       Ord_20   Prod_1    SHP_27    370.48      0.04              28   
2        Ord_5   Prod_1     SHP_7    905.08      0.09              22   
3        Ord_6   Prod_1     SHP_8   2781.82      0.07              21   
4       Ord_55   Prod_1    SHP_72   3338.98      0.00              40   
...        ...      ...       ...       ...       ...             ...   
8394  Ord_5400  Prod_17  SHP_7540   5793.46      0.01              38   
8395  Ord_5450  Prod_17  SHP_7690  12007.05      0.09              16   
8396  Ord_5485  Prod_17  SHP_7664   4233.15      0.08              35   
8397  Ord_5489  Prod_17  SHP_7671    754.92      0.00               7   
8398  Ord_5504  Prod_17  SHP_7696  10071.09      0.10              41   

       Profit  Shipping_Cost  Product_Base_Margin Product_Category  \
0     -213.25          35.00                 0.80  OFFICE SUPPLIES   
1       -5.45           4.51                 0.59  OFFICE SUPPLIES   
2      127.70           6.22                  NaN  OFFICE SUPPLIES   
3     -695.26          35.00                  NaN  OFFICE SUPPLIES   
4     -846.73          35.00                 0.81  OFFICE SUPPLIES   
...       ...            ...                  ...              ...   
8394  2477.77          13.99                 0.38       TECHNOLOGY   
8395  2713.95          55.30                 0.40       TECHNOLOGY   
8396  1219.87          26.30                 0.38       TECHNOLOGY   
8397  -129.57          19.99                 0.52       TECHNOLOGY   
8398  1977.69          17.86                 0.58       TECHNOLOGY   

        Product_Sub_Category  
0     STORAGE & ORGANIZATION  
1     STORAGE & ORGANIZATION  
2     STORAGE & ORGANIZATION  
3     STORAGE & ORGANIZATION  
4     STORAGE & ORGANIZATION  
...                      ...  
8394         OFFICE MACHINES  
8395         OFFICE MACHINES  
8396         OFFICE MACHINES  
8397         OFFICE MACHINES  
8398         OFFICE MACHINES  

[8399 rows x 16 columns]

In [177]:
# Select all orders from product category = office supplies and from the corporate segment
office_corporate = df2.loc[(df2["Product_Category"]=="OFFICE SUPPLIES")&(df2["Customer_Segment"]=='CORPORATE')]
office_corporate

Customer_Name               Province                 Region  \
2       CARL JACKSON                NUNAVUT                NUNAVUT   
3     MONICA FEDERLE                NUNAVUT                NUNAVUT   
6        BRYAN DAVIS  NORTHWEST TERRITORIES  NORTHWEST TERRITORIES   
7        ALAN BARNES  NORTHWEST TERRITORIES  NORTHWEST TERRITORIES   
8       JAMIE KUNITZ  NORTHWEST TERRITORIES  NORTHWEST TERRITORIES   
...              ...                    ...                    ...   
8050     CHUCK CLARK                ALBERTA                   WEST   
8052   MARIA ZETTNER                ALBERTA                   WEST   
8057  KELLY WILLIAMS                ALBERTA                   WEST   
8058     RUBEN DARTT                ALBERTA                   WEST   
8060   AARON BERGMAN                ALBERTA                   WEST   

     Customer_Segment    Cust_id    Ord_id  Prod_id   Ship_id    Sales  \
2           CORPORATE     Cust_5     Ord_5   Prod_1     SHP_7   905.08   
3           CORPORATE     Cust_6     Ord_6   Prod_1     SHP_8  2781.82   
6           CORPORATE    Cust_43    Ord_58   Prod_1    SHP_77  2039.56   
7           CORPORATE    Cust_45   Ord_209   Prod_1   SHP_286   127.80   
8           CORPORATE    Cust_50    Ord_75   Prod_1    SHP_99    73.07   
...               ...        ...       ...      ...       ...      ...   
8050        CORPORATE  Cust_1720  Ord_5062  Prod_16  SHP_7068   587.71   
8052        CORPORATE  Cust_1740  Ord_5135  Prod_16  SHP_7176    20.19   
8057        CORPORATE  Cust_1774   Ord_650  Prod_16  SHP_7421    33.54   
8058        CORPORATE  Cust_1793  Ord_5344  Prod_16  SHP_7462     6.93   
8060        CORPORATE  Cust_1818  Ord_5446  Prod_16  SHP_7609   136.81   

      Discount  Order_Quantity  Profit  Shipping_Cost  Product_Base_Margin  \
2         0.09              22  127.70           6.22                  NaN   
3         0.07              21 -695.26          35.00                  NaN   
6         0.04               7 -329.49          35.00                 0.80   
7         0.07               8  -30.48           7.51                 0.57   
8         0.08               4  -41.01          11.25                 0.60   
...        ...             ...     ...            ...                  ...   
8050      0.09              49   26.28           3.14                 0.60   
8052      0.04               6  -13.44           1.92                 0.84   
8057      0.08               4  -29.49           8.37                 0.58   
8058      0.01               2   -4.64           2.56                 0.55   
8060      0.01              23  -30.51           3.60                 0.56   

     Product_Category           Product_Sub_Category  
2     OFFICE SUPPLIES         STORAGE & ORGANIZATION  
3     OFFICE SUPPLIES         STORAGE & ORGANIZATION  
6     OFFICE SUPPLIES         STORAGE & ORGANIZATION  
7     OFFICE SUPPLIES         STORAGE & ORGANIZATION  
8     OFFICE SUPPLIES         STORAGE & ORGANIZATION  
...               ...                            ...  
8050  OFFICE SUPPLIES  SCISSORS, RULERS AND TRIMMERS  
8052  OFFICE SUPPLIES  SCISSORS, RULERS AND TRIMMERS  
8057  OFFICE SUPPLIES  SCISSORS, RULERS AND TRIMMERS  
8058  OFFICE SUPPLIES  SCISSORS, RULERS AND TRIMMERS  
8060  OFFICE SUPPLIES  SCISSORS, RULERS AND TRIMMERS  

[1680 rows x 16 columns]

Similary, you can merge the other dimension tables - shipping_df and orders_df to create a master_df and perform indexing using any column in the master dataframe.

In [179]:
# Merging shipping_df store result in df3
df3 = pd.merge(df2,shipping_df)
df3

Customer_Name Province   Region Customer_Segment    Cust_id  \
0     MUHAMMED MACINTYRE  NUNAVUT  NUNAVUT   SMALL BUSINESS     Cust_1   
1         CARLOS SOLTERO  NUNAVUT  NUNAVUT         CONSUMER     Cust_4   
2           CARL JACKSON  NUNAVUT  NUNAVUT        CORPORATE     Cust_5   
3         MONICA FEDERLE  NUNAVUT  NUNAVUT        CORPORATE     Cust_6   
4          NICOLE HANSEN  NUNAVUT  NUNAVUT   SMALL BUSINESS    Cust_24   
...                  ...      ...      ...              ...        ...   
8394            SHUI TOM  ALBERTA     WEST      HOME OFFICE  Cust_1816   
8395        FRANK HAWLEY  ALBERTA     WEST      HOME OFFICE  Cust_1817   
8396       AARON BERGMAN  ALBERTA     WEST        CORPORATE  Cust_1818   
8397        ADRIAN SHAMI  ALBERTA     WEST         CONSUMER  Cust_1820   
8398        HARRY GREENE  ALBERTA     WEST        CORPORATE  Cust_1830   

        Ord_id  Prod_id   Ship_id     Sales  Discount  Order_Quantity  \
0        Ord_1   Prod_1     SHP_1    261.54      0.04               6   
1       Ord_20   Prod_1    SHP_27    370.48      0.04              28   
2        Ord_5   Prod_1     SHP_7    905.08      0.09              22   
3        Ord_6   Prod_1     SHP_8   2781.82      0.07              21   
4       Ord_55   Prod_1    SHP_72   3338.98      0.00              40   
...        ...      ...       ...       ...       ...             ...   
8394  Ord_5400  Prod_17  SHP_7540   5793.46      0.01              38   
8395  Ord_5450  Prod_17  SHP_7690  12007.05      0.09              16   
8396  Ord_5485  Prod_17  SHP_7664   4233.15      0.08              35   
8397  Ord_5489  Prod_17  SHP_7671    754.92      0.00               7   
8398  Ord_5504  Prod_17  SHP_7696  10071.09      0.10              41   

       Profit  Shipping_Cost  Product_Base_Margin Product_Category  \
0     -213.25          35.00                 0.80  OFFICE SUPPLIES   
1       -5.45           4.51                 0.59  OFFICE SUPPLIES   
2      127.70           6.22                  NaN  OFFICE SUPPLIES   
3     -695.26          35.00                  NaN  OFFICE SUPPLIES   
4     -846.73          35.00                 0.81  OFFICE SUPPLIES   
...       ...            ...                  ...              ...   
8394  2477.77          13.99                 0.38       TECHNOLOGY   
8395  2713.95          55.30                 0.40       TECHNOLOGY   
8396  1219.87          26.30                 0.38       TECHNOLOGY   
8397  -129.57          19.99                 0.52       TECHNOLOGY   
8398  1977.69          17.86                 0.58       TECHNOLOGY   

        Product_Sub_Category  Order_ID       Ship_Mode   Ship_Date  
0     STORAGE & ORGANIZATION         3     REGULAR AIR  20-10-2010  
1     STORAGE & ORGANIZATION      1792     REGULAR AIR  13-11-2010  
2     STORAGE & ORGANIZATION       613     REGULAR AIR  18-06-2011  
3     STORAGE & ORGANIZATION       643     EXPRESS AIR  25-03-2011  
4     STORAGE & ORGANIZATION      8833     REGULAR AIR  08-05-2012  
...                      ...       ...             ...         ...  
8394         OFFICE MACHINES     17287     REGULAR AIR  06-04-2011  
8395         OFFICE MACHINES     36772  DELIVERY TRUCK  17-05-2010  
8396         OFFICE MACHINES     54019  DELIVERY TRUCK  08-07-2009  
8397         OFFICE MACHINES     56706     REGULAR AIR  11-11-2009  
8398         OFFICE MACHINES     54279  DELIVERY TRUCK  31-07-2011  

[8399 rows x 19 columns]

In [245]:
# Merging the orders table to create a master_df
master_df = pd.merge(df3,orders_df)
master_df.head()

Customer_Name Province   Region Customer_Segment Cust_id  Ord_id  \
0  MUHAMMED MACINTYRE  NUNAVUT  NUNAVUT   SMALL BUSINESS  Cust_1   Ord_1   
1      CARLOS SOLTERO  NUNAVUT  NUNAVUT         CONSUMER  Cust_4  Ord_20   
2        CARL JACKSON  NUNAVUT  NUNAVUT        CORPORATE  Cust_5   Ord_5   
3        CARL JACKSON  NUNAVUT  NUNAVUT        CORPORATE  Cust_5   Ord_5   
4      MONICA FEDERLE  NUNAVUT  NUNAVUT        CORPORATE  Cust_6   Ord_6   

  Prod_id Ship_id    Sales  Discount  ...  Profit  Shipping_Cost  \
0  Prod_1   SHP_1   261.54      0.04  ... -213.25          35.00   
1  Prod_1  SHP_27   370.48      0.04  ...   -5.45           4.51   
2  Prod_1   SHP_7   905.08      0.09  ...  127.70           6.22   
3  Prod_3   SHP_6    93.54      0.03  ...  -54.04           7.72   
4  Prod_1   SHP_8  2781.82      0.07  ... -695.26          35.00   

   Product_Base_Margin  Product_Category            Product_Sub_Category  \
0                 0.80   OFFICE SUPPLIES          STORAGE & ORGANIZATION   
1                 0.59   OFFICE SUPPLIES          STORAGE & ORGANIZATION   
2                  NaN   OFFICE SUPPLIES          STORAGE & ORGANIZATION   
3                 0.38   OFFICE SUPPLIES  BINDERS AND BINDER ACCESSORIES   
4                  NaN   OFFICE SUPPLIES          STORAGE & ORGANIZATION   

  Order_ID    Ship_Mode   Ship_Date  Order_Date Order_Priority  
0        3  REGULAR AIR  20-10-2010  13-10-2010            LOW  
1     1792  REGULAR AIR  13-11-2010  08-11-2010            LOW  
2      613  REGULAR AIR  18-06-2011  17-06-2011           HIGH  
3      613  REGULAR AIR  17-06-2011  17-06-2011           HIGH  
4      643  EXPRESS AIR  25-03-2011  24-03-2011           HIGH  

[5 rows x 21 columns]

In [316]:
master_df.shape

(8399, 22)

In [246]:
master_df["Profit"].min()

-14140.7

In [306]:
# Which customer segments are the least profitable? 
# Use grouping with master_df on Customer_Segment
cust_group = master_df.groupby("Customer_Segment")["Profit"].sum()
print(cust_group)
cust_group.idxmin()

Customer_Segment
CONSUMER          287961.0
CORPORATE         599748.2
HOME OFFICE       318354.8
SMALL BUSINESS    315708.4
Name: Profit, dtype: float64


'CONSUMER'

In [307]:
# Which product categories are the least profitable?
prod_group = master_df.groupby("Product_Category")["Profit"].sum()
print(prod_group)
prod_group.idxmin()

Product_Category
FURNITURE          117432.9
OFFICE SUPPLIES    518023.6
TECHNOLOGY         886315.9
Name: Profit, dtype: float64


'FURNITURE'

In [308]:
# Which product categories and sub-categories are the least profitable?
prod_cat_subcat = master_df.groupby(["Product_Category","Product_Sub_Category"])["Profit"].sum()
print(prod_cat_subcat)
prod_cat_subcat.idxmin()

Product_Category  Product_Sub_Category          
FURNITURE         BOOKCASES                         -33582.8
                  CHAIRS & CHAIRMATS                149649.6
                  OFFICE FURNISHINGS                100427.8
                  TABLES                            -99061.7
OFFICE SUPPLIES   APPLIANCES                         97159.2
                  BINDERS AND BINDER ACCESSORIES    307414.5
                  ENVELOPES                          48182.0
                  LABELS                             13677.4
                  PAPER                              45262.2
                  PENS & ART SUPPLIES                 7566.2
                  RUBBER BANDS                        -102.2
                  SCISSORS, RULERS AND TRIMMERS      -7798.7
                  STORAGE & ORGANIZATION              6663.0
TECHNOLOGY        COMPUTER PERIPHERALS               94288.3
                  COPIERS AND FAX                   167361.3
                  OFFICE MACHINES   

('FURNITURE', 'TABLES')

In [190]:
master_df.columns

Index(['Customer_Name', 'Province', 'Region', 'Customer_Segment', 'Cust_id',
       'Ord_id', 'Prod_id', 'Ship_id', 'Sales', 'Discount', 'Order_Quantity',
       'Profit', 'Shipping_Cost', 'Product_Base_Margin', 'Product_Category',
       'Product_Sub_Category', 'Order_ID', 'Ship_Mode', 'Ship_Date',
       'Order_Date', 'Order_Priority'],
      dtype='object')

In [191]:
# Customers in which geographic region are the least profitable?
master_df.groupby("Region")["Profit"].min()

Region
ATLANTIC                -11769.17
NORTHWEST TERRITORIES   -11053.60
NUNAVUT                  -1766.01
ONTARIO                  -6474.65
PRARIE                  -11861.46
QUEBEC                   -9611.91
WEST                    -14140.70
YUKON                    -9078.94
Name: Profit, dtype: float64

In [208]:
help(master_df.agg)

Help on method aggregate in module pandas.core.frame:

aggregate(func, axis=0, *args, **kwargs) method of pandas.core.frame.DataFrame instance
    Aggregate using one or more operations over the specified axis.
    
    .. versionadded:: 0.20.0
    
    Parameters
    ----------
    func : function, str, list or dict
        Function to use for aggregating the data. If a function, must either
        work when passed a DataFrame or when passed to DataFrame.apply.
    
        Accepted combinations are:
    
        - function
        - string function name
        - list of functions and/or function names, e.g. ``[np.sum, 'mean']``
        - dict of axis labels -> functions, function names or list of such.
    axis : {0 or 'index', 1 or 'columns'}, default 0
            If 0 or 'index': apply function to each column.
            If 1 or 'columns': apply function to each row.
    *args
        Positional arguments to pass to `func`.
    **kwargs
        Keyword arguments to pass to `fun

In [314]:
# Calculate the Sales across each region as a percentage of total Sales
# hint -the resulting object is a Series, thus you can perform vectorised computations on them
# You can divide the entire series by a number (total sales) easily 
total_sales = master_df["Sales"].sum()
grp_sales_agg = master_df.groupby('Region').agg({'Sales':'sum'})
grp_sales_agg/total_sales

Sales
Region                         
ATLANTIC               0.135043
NORTHWEST TERRITORIES  0.053692
NUNAVUT                0.007802
ONTARIO                0.205370
PRARIE                 0.190224
QUEBEC                 0.101249
WEST                   0.241194
YUKON                  0.065426

In [239]:
# Create a new column is_profitable
# The value in the coulmn is_profitable is True when the profit >0  else False
# Do the work using apply
master_df["is_profitable"] = master_df["Profit"].apply(lambda x: x>0)

In [240]:
master_df.columns

Index(['Customer_Name', 'Province', 'Region', 'Customer_Segment', 'Cust_id',
       'Ord_id', 'Prod_id', 'Ship_id', 'Sales', 'Discount', 'Order_Quantity',
       'Profit', 'Shipping_Cost', 'Product_Base_Margin', 'Product_Category',
       'Product_Sub_Category', 'Order_ID', 'Ship_Mode', 'Ship_Date',
       'Order_Date', 'Order_Priority', 'is_profitable'],
      dtype='object')

In [241]:
master_df["is_profitable"]

0       False
1       False
2        True
3       False
4       False
        ...  
8394     True
8395    False
8396     True
8397    False
8398     True
Name: is_profitable, Length: 8399, dtype: bool

In [250]:
# Comparing percentage of profitable orders across product categories using groupby
product_percent = master_df.groupby('Product_Category')["Profit"]

In [251]:
# you want to see Profit as one decimal place
# apply the round() function 
# the first value is -30.5
# the first value before is -30.51
# Modify the profit column itself
def profit_round(x):
    return round(x,1)
master_df['Profit'] = master_df['Profit'].apply(profit_round)
master_df['Profit'].head()

0   -213.2
1     -5.5
2    127.7
3    -54.0
4   -695.3
Name: Profit, dtype: float64

In [252]:
# Creating a column Profit / Order_Quantity
# The new columnname is 'profit_per_qty'
master_df["profit_per_day"] = master_df["Profit"]/master_df["Order_Quantity"]
master_df["profit_per_day"]

0        -35.533333
1         -0.196429
2          5.804545
3         -4.500000
4        -33.109524
           ...     
8394     169.008333
8395   -3974.500000
8396      34.854286
8397     -18.514286
8398      48.236585
Name: profit_per_day, Length: 8399, dtype: float64

In [256]:
# Compare average Sales across customer segments using pivot table
pd.pivot_table(master_df,index="Customer_Segment",values="Sales")

Sales
Customer_Segment             
CONSUMER          1857.859965
CORPORATE         1787.680389
HOME OFFICE       1754.312931
SMALL BUSINESS    1698.124841

In [257]:
# compare total number of profitable orders across regions using pivot_table
# Note that since is_profitable is 1/0, we can directly compute the sum
pd.pivot_table(master_df,index="Region",values="Profit", aggfunc=['count', 'mean'])

count        mean
                      Profit      Profit
Region                                  
ATLANTIC                1080  221.261111
NORTHWEST TERRITORIES    394  255.464467
NUNAVUT                   79   35.965823
ONTARIO                 1826  189.960953
PRARIE                  1706  188.254513
QUEBEC                   781  179.803073
WEST                    1991  149.175892
YUKON                    542  136.254244

In [258]:
# Compare the total profit across product categories and customer segments
pd.pivot_table(master_df,index=["Product_Category","Customer_Segment"],values="Profit", aggfunc=['count', 'mean'])

count        mean
                                  Profit      Profit
Product_Category Customer_Segment                   
FURNITURE        CONSUMER            358  119.353352
                 CORPORATE           633   34.768246
                 HOME OFFICE         406   59.060591
                 SMALL BUSINESS      327   87.821101
OFFICE SUPPLIES  CONSUMER            865  102.349595
                 CORPORATE          1680  120.856429
                 HOME OFFICE        1145  105.804017
                 SMALL BUSINESS      920  114.463913
TECHNOLOGY       CONSUMER            426  367.840610
                 CORPORATE           763  491.089253
                 HOME OFFICE         481  360.146778
                 SMALL BUSINESS      395  459.959747